client-lib/nsa.js

```diff
@@ -0,0 +1,210 @@
+  // a work around for a bug in the audiosource dependency
+  if (opts && opts.replay) {
+    playerMap[window.AppData.player].play(0.001);
+  } else playerMap[window.AppData.player].play();
+
+  window.AppData.playing = true;
+  currentStep = requestAnimationFrame(step);
+});
+
+emitter.on('pause', (opts) => {
+  sendMetricsEvent('player_view', 'pause');
+  playerMap[window.AppData.player].pause();
+  window.AppData.playing = false;
+});
+
+emitter.on('mute', (opts) => {
+  sendMetricsEvent('player_view', 'mute');
+  playerMap[window.AppData.player].mute();
+  window.AppData = deepAssign(window.AppData, {muted: true});
+});
+
+emitter.on('unmute', (opts) => {
+  sendMetricsEvent('player_view', 'unmute');
+  playerMap[window.AppData.player].unmute();
+  window.AppData = deepAssign(window.AppData, {muted: false});
+});
+
+emitter.on('replay', (opts) => {
+  sendMetricsEvent('player_view', 'replay');
+  emitter.emit('reset');
+  emitter.emit('play', {replay: true});
+});
+
+emitter.on('load', (opts) => {
+  sendMetricsEvent('player_view', 'video_loaded');
+
+  // initial step to set times
+  step();
+
+  window.AppData = deepAssign(window.AppData, {
+    loaded: true,
+    duration: opts.duration
+  });
+});
+
+emitter.on('set-volume', (opts) => {
+  playerMap[window.AppData.player].volume = opts.value;
+  window.AppData = deepAssign(window.AppData, {
+    volume: opts.value
+  });
+});
+
+emitter.on('set-time', (opts) => {
+  playerMap[window.AppData.player].time = opts.value;
+
+  // if we are paused force the ui to update
+  if (!window.AppData.playing) {
+    window.AppData = deepAssign(window.AppData, {
+      time: `${formatTime(opts.value)} / ${formatTime(window.AppData.duration)}`,
+      progress: opts.value / window.AppData.duration,
+      currentTime: opts.value
+    });
+  }
+});
+
+emitter.on('update-visual', (opts) => {
+  if (window.AppData.player !== 'audio') return;
+  if (playerMap[window.AppData.player].fft.type === 'time') {
+    playerMap[window.AppData.player].fft.type = 'frequency';
+  } else playerMap[window.AppData.player].fft.type = 'time';
+});
+
+emitter.on('resize', (opts) => {
+  if (window.AppData.player !== 'audio') return;
+  const spacing = (opts.width < 1000) ? 1 : opts.width / 500;
+  playerMap[window.AppData.player].fft.width = spacing;
+  playerMap[window.AppData.player].fft.spacing = spacing;
+});
+
+emitter.on('close', () => {
+  sendMetricsEvent(getView(), 'close');
+  if (playerMap[window.AppData.player]) {
+    playerMap[window.AppData.player].remove();
+  }
+  playerMap = {};
+  sendToAddon({action: 'close'});
+  window.AppData.error = false;
+});
+
+emitter.on('minimize', () => {
+  sendMetricsEvent(getView(), 'minimize');
+  sendToAddon({action: 'minimize'});
+  window.AppData.minimized = true;
+});
+
+emitter.on('maximize', () => {
+  sendMetricsEvent(getView(), 'maximize');
+  sendToAddon({action: 'maximize'});
+  window.AppData.minimized = false;
+});
+
+emitter.on('send-to-tab', () => {
```

Comment:
or, at least create another listener that does the sendToAddon bit. might require another event, since it looks like currentTime gets updated just before the addon is notified.